### DISCUSSION TITLE
Pico W STA/AP Interfaces Sharing Same DNS?

### DISCUSSION THREAD

**Discussion by Beormund** [micropython/micropython#11295] [RP2040 / Pico]
The following is causing havoc with my AP captive portal:

```py
# Start clean: run machine.reset()
# import machine
# machine.reset()
# MicroPython v1.19.1-1016-gb525f1c9e on 2023-04-14; Raspberry Pi Pico W with RP2040

import network

wap = network.WLAN(network.AP_IF)
print(f'Default Inactive AP: {wap.ifconfig()}')
wap.active(True)
print(f'Default Active AP: {wap.ifconfig()}')
# Update AP_IF DNS
wap.ifconfig(wap.ifconfig()[:3] + (wap.ifconfig()[0],))
print(f'Updated Active AP: {wap.ifconfig()}')

sta = network.WLAN(network.STA_IF)
print(f'Default Inactive STA: {sta.ifconfig()}')
# After sta.connect(ssid, pwd), DNS is updated:
# Mimic DNS update:
sta.ifconfig(sta.ifconfig()[:3] + ('192.168.1.254',))
print(f'STA After STA DNS Update: {sta.ifconfig()}')

# Why is AP_IF DNS also updated?
print(f'AP After STA DNS Update: {wap.ifconfig()}')

#>>> %Run -c $EDITOR_CONTENT
# Default Inactive AP: ('0.0.0.0', '0.0.0.0', '0.0.0.0', '0.0.0.0')
# Default Active AP: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '0.0.0.0')
# Updated Active AP: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '192.168.4.1')
# Default Inactive STA: ('0.0.0.0', '0.0.0.0', '0.0.0.0', '192.168.4.1')
# STA After STA DNS Update: ('0.0.0.0', '0.0.0.0', '0.0.0.0', '192.168.1.254')
# AP After STA DNS Update: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '192.168.1.254')
#>>> 
```
In my application I run a captive portal using the Pico W's soft access point, along with a lightweight DNS server. This enables a user to select a local network and enter a password. This all works well until the STA_IF interface connects and gets an IP. I should be able to display this IP in the captive portal so the user can then copy it and navigate to the new IP once the Pico W has restarted. Unfortunately the captive portal stops working because the AP_IF DNS changes.

I don't think there is a correct separation of DNS between the two interfaces.

**Comment by karfas** [comment]
> I don't think there is a correct separation of DNS between the two interfaces.

Why should each Interface use it's own DNS server ? Even in Linux this is not possible out-of-the-box (see [Link](https://serverfault.com/questions/423882/configure-a-dns-server-per-nic-interface-eth0-eth1)) - this would require some firewall/iptables magic or special config entries I'm not aware of.

**Reply by Beormund** [reply]
> > I don't think there is a correct separation of DNS between the two interfaces.
> 
> Why should each Interface use it's own DNS server ? Even in Linux this is not possible out-of-the-box (see [Link](https://serverfault.com/questions/423882/configure-a-dns-server-per-nic-interface-eth0-eth1)) - this would require some firewall/iptables magic or special config entries I'm not aware of.

Because a soft access point can provide its own dns server to redirect traffic to its own web server - that's how a captive portal works. In station  mode the dns server can be anywhere on  the network or upstream.

**Reply by karfas** [reply]
It's not a requirement that the AP of a captive portal provides it's own name server. Redirection of traffic to the portal's http server can be done by other means (not provided by stock micropython, anyway).

However, any AP will provide a configured DNS server address to connected clients.

In your case, you decided to (implicit) **configure** a DNS server address by using the DHCP of the client connection. Just use a static address for your STA connection and provide the wanted DNS address.

**Comment by vinux84** [comment]
Resurrecting this post, just curious @Beormund , were you ever able to fix the issue?..I'm confronting a similar problem right now

**Reply by Beormund** [reply]
Unfortunately not. I gave up on the RP Pico W due to lack of RAM for what I needed. I Switched to ESP32. May take another look when the next version of the RP2040 comes out.

**Reply by samveen** [reply]
@Beormund I'm facing a similar issue too (discussion at https://forums.raspberrypi.com/viewtopic.php?t=347699).

Do you have the DNS server code you're using listed somewhere? What I'm trying to do is as follows:
- Set up the the STA.
- Set up the IF
- run a micro dns-resolver that manually recurses the requests to the DNS server configured on the IF. This is where 
I am failing.

Maybe I'll be able to figure out where my failure happens.

**Comment by vinux84** [comment]
thanks for the response.

**Comment by vinux84** [comment]
@samveen - Looks like we have very similar issues. All I'm wanting is for the pico to advertise its host name using mDNS or some other way,  no matter what router it connects too, it needs to be consistent resolving its DNS name on what ever router it hooks up to.  (SOHO router or even enterprise/business)

From what I gather for the most recent micropython runtime, mDNS issues have been solved. But the issue I'm having is there is hardly any documentation on mDNS for micropython. And I'm not really sure where to start, I have read post where you have to enable it, but again not sure on how to go about that. 

- I have been able to assign it a hostname  - network.hostname("hostname") , and indeed I will get my hostname for the pico show when I search my network. But the problem I'm having 

- I can't access the webpage the pico serves up using a url, using: hostname.local 

- Oddly enough I can ping it using just hostname. 

This leads me to believe mDNS still isn't working because .local doesn't work. You may know more then me, Am I missing something here? 

If not my search continues for the pico to advertise itself

**Reply by vinux84** [reply]
update: looks like I have found a way via this page > https://github.com/arag0re/micropython/releases/tag/mDNS-fixed

**Reply by samveen** [reply]
@vinux84 Looks like that repository disables the mDNS responder for lwip  (https://github.com/arag0re/micropython/commit/ba3ef932a25dde522b1f6abec93fe9a3bb8ff646)

My problem isn't mDNS. My issue is that I have 2 links; let me name them as `wifi` (as `AP_IF`) and `wan`(as `STA_IF`).  `wifi` is an `AP` where clients connect. `wan` is an `STA` that connects to another WiFi as a client. I need to take the DNS requests that arrive on the Pico through the `wifi` interface, and then send them to the DNS server on the `wan` link.

**Reply by Sagarhs45** [reply]
@samveen @Beormund @vinux84 @karfas If anyone can help me in implementing captive portal in CY8CKIT-062S2-43012 using lwip stack, please let me know. It would be really helpful

**Reply by samveen** [reply]
@Sagarhs45 I'm still stuck where I was, with my project in limbo. It's just a personal  What-If project, so not much damage, luckily.

**Reply by Sagarhs45** [reply]
@samveen Do you have idea of setting up of local dns server using lwip?
